src/control/cmd/daos_agent/procmon.go

```diff
@@ -135,56 +141,115 @@ func NewProcMon(logger logging.Logger) *procMon {
 	}
 }
 
-func (p *procMon) RegisterProcess(ctx context.Context, Pid int32) {
-	p.submitRequest(ctx, Pid, drpc.MethodGetAttachInfo)
+func (p *procMon) RegisterPool(ctx context.Context, Pid int32, poolReq *mgmtpb.PoolMonitorReq) {
+	req := &procMonRequest{
+		pid:            Pid,
+		action:         drpc.MethodPoolConnect,
+		poolUUID:       poolReq.PoolUUID,
+		poolHandleUUID: poolReq.PoolHandleUUID,
+	}
+	p.submitRequest(ctx, req)
 }
 
-func (p *procMon) UnregisterProcess(ctx context.Context, Pid int32) {
-	p.submitRequest(ctx, Pid, drpc.MethodDisconnect)
+func (p *procMon) DisconnectPool(ctx context.Context, Pid int32, poolReq *mgmtpb.PoolMonitorReq) {
+	req := &procMonRequest{
+		pid:            Pid,
+		action:         drpc.MethodPoolDisconnect,
+		poolUUID:       poolReq.PoolUUID,
+		poolHandleUUID: poolReq.PoolHandleUUID,
+	}
+	p.submitRequest(ctx, req)
 }
 
-func (p *procMon) submitRequest(ctx context.Context, Pid int32, Action drpc.MgmtMethod) {
+func (p *procMon) DisconnectProcess(ctx context.Context, Pid int32) {
 	req := &procMonRequest{
 		pid:    Pid,
-		action: Action,
+		action: drpc.MethodDisconnect,
 	}
+	p.submitRequest(ctx, req)
+}
+
+func (p *procMon) submitRequest(ctx context.Context, request *procMonRequest) {
 	select {
 	case <-ctx.Done():
-	case p.request <- req:
+	case p.request <- request:
 	}
 }
 
-func (p *procMon) handleProcessAttach(ctx context.Context, request *procMonRequest) {
+func (p *procMon) handlePoolConnect(ctx context.Context, request *procMonRequest) {
+	p.log.Debugf("Received request to connect pool:%s with handle %s, for pid:%d", request.poolUUID, request.poolHandleUUID, request.pid)
+
 	info, found := p.procs[request.pid]
 
-	if found {
-		p.log.Errorf("Attempted to monitor process %d more than once", request.pid)
+	if !found {
+		child, cancel := context.WithCancel(ctx)
+		info = &procInfo{
+			log:       p.log,
+			pid:       request.pid,
+			cancelCtx: cancel,
+			response:  p.response,
+			handles:   make(map[string]map[string]bool),
+		}
+
+		p.procs[request.pid] = info
+		go info.monitorProcess(child)
+	}
+
+	_, found = info.handles[request.poolUUID]
+	if !found {
+		info.handles[request.poolUUID] = make(map[string]bool)
+	}
+	info.handles[request.poolUUID][request.poolHandleUUID] = true
+
+}
+
+func (p *procMon) handlePoolDisconnect(request *procMonRequest) {
```

Comment:
I think you will want mutex protection while inserting and removing elements from the maps to avoid race conditions in this function, and the others where you touch it.

Reply:
I don't believe this is necessary since requests are serialized by the go channels. No matter how many requests come in there shouldn't ever be two people accessing this data structure at the same time.